Add removal of client label

danielberbece opened the pull request:
Added a feature to let an user remove an already assigned label of a client. The feature consists of an `X` button inside the client label chip. Upon click, a POST request is sent to the endpoint `/api/v2/clients/labels/remove` with the label name and the clientId of the selected client. After the POST request, the client is re-fetched and the UI is automatically updated in order to reflect the change (be it successful or not).

panhania commented on grr/server/grr_response_server/gui/ui/components/client/client.ts:
```diff
@@ -49,6 +49,10 @@ export class Client implements OnInit, OnDestroy {
     });
   }
 
+  removeLabel(labelName: string) {
+    this.clientPageFacade.removeClientLabel(labelName);
+  }
```
Nit: the argument name in `addLabel` (and pretty much everything else) is `label`, but here it is `labelName`—looks a bit inconsistent.

danielberbece replied:
True, will change to `label`

max-vogler commented:
This looks like a good PR for @panhania to review! I suggest displaying a snackbar with an undo button after a label has been removed, see [Material Design: Confirmation & acknowledgement](https://material.io/design/communication/confirmation-acknowledgement.html#acknowledgement).

danielberbece commented:
I addressed the issue and added the suggested feature: a snackbar with the message `Label <name> removed` and action `UNDO` when the request to remove the label succeeds.

panhania commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -491,7 +503,13 @@ export class ClientPageFacade {
   }
+
+  /** Removes a label from the selected client */
+  removeClientLabelReq(label: string): Observable<HttpResponse<{}>> {
+    return this.store.removeClientLabel(label);
+  }
```
I find it a bit awkward that we have a `addClientLabel` method (that returns nothing) and a `removeClientLabelReq` method (that returns an observable). Is this inconsistency necessary?

danielberbece replied:
Totally agree here, but in order to show the `UNDO` snackbar we needed to know if the remove request is successful or not. If there is a better way to achieve this behaviour, I am open to suggestions.

panhania replied:
@max-vogler should have some suggestions.

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client/client.ng.html:
```diff
@@ -7,8 +7,8 @@ <h1 class="mat-h1">{{client.knowledgeBase.fqdn}}</h1>
           <online-chip [lastSeen]="client.lastSeenAt"></online-chip>
-          <mat-chip *ngFor="let label of client.labels">
-            {{label.name}}
+          <mat-chip *ngFor="let label of client.labels" (removed)="removeLabel(label.name)">
```
Please add `trackBy` and track elements by label name. Two objects are never `===` unless they are the same instance, so Angular will recreate the DOM upon every change even though labels technically stay the same.

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api/http_api_service.ts:
```diff
@@ -223,6 +223,14 @@ export class HttpApiService {
         url, {client_ids: [clientId], labels: [label]});
   }
 
+  removeClientLabel(clientId: string, label: string): Observable<HttpResponse<{}>> {
```
HttpApiResponse is an implementation detail and should not be returned. Instead, return {} or void.

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api/http_api_service.ts:
```diff
@@ -223,6 +223,14 @@ export class HttpApiService {
+    const url = `${URL_PREFIX}/clients/labels/remove`;
+    return this.http.post<{}>(
+        url, {client_ids: [clientId], labels: [label]}, {
+          observe: 'response',
```
Im on the run right now, cannot look it up: What does observe response do?

danielberbece replied:
By default, the returned observable would be the data we are requesting for. In these case, I wanted to get the status, and in order to do that I had to specify `response` to be returned.

max-vogler commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -428,6 +429,17 @@ export class ClientPageStore extends ComponentStore<ClientPageState> {
           ));
+
+  // An effect to remove a label from the selected client
+  removeClientLabel(label: string): Observable<HttpResponse<{}>> {
```
Functions that trigger change should only return void in NgRx (our store implementation). Instead of returning an observable here, create an observable as member of the Store. Then, when deletion of the label is completed, update the state which should then trigger emission of the dedicated observable. E.g.: the observable that tracks flow creation state.

max-vogler commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -491,7 +503,13 @@ export class ClientPageFacade {
   }
+
+  /** Removes a label from the selected client */
+  removeClientLabelReq(label: string): Observable<HttpResponse<{}>> {
```
`Req` sounds like an implementation detail. Please name it `removeClientLabel`.

max-vogler reviewed (changes requested):
Please add tests for the undo function.

max-vogler commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -32,6 +33,17 @@ export type StartFlowState = {
   readonly error: string,
 };
 
+export type ChangeRequestState = {
```
This can be simplified and `clearRemoveClientLabelState` removed:

We are only interested in the fact that a label was deleted successfully (to show the snackbar and remove it from the label list). You can introduce an observable that only emits the label name whenever the deletion was successful.

danielberbece replied:
This is a very good idea, thanks for the suggestion!

danielberbece commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -457,6 +458,27 @@ export class ClientPageStore extends ComponentStore<ClientPageState> {
+          switchMap(
+              ([label, clientId]) =>
+                  this.httpApiService.removeClientLabel(clientId, label)),
+          catchError(() => {
```
Removing the `catchError()` pipe step causes the store `effect` to throw the error. I'd suggest to catch the error here in order to throw it further in the `removedClientLabels$` observable

max-vogler commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -457,6 +458,26 @@ export class ClientPageStore extends ComponentStore<ClientPageState> {
+                  this.httpApiService.removeClientLabel(clientId, label)),
+          tap((label) => {
+            this.fetchClient();
+            this.removedClientLabels$.next(label);
```
Please loop this through the store's state: Extend `ClientPageState` with a field (e.g. lastRemovedClientLabel`), which is then updated on successful removal (in the `tap`). You can ignore the errors for now - remove `catchError`.

removedClientLabels$ should not be a `new Subject`, instead, use `this.selectt()` to read the property of the state.

Looping all state changes through one state is a typical pattern for NgRx.
